File: LSDALTON/tools/xyz2mol.pl
#!/usr/bin/perl -w
# A script for converting xyz input (Gaussian-type) to dalton mol input.
# First two lines are treated as a comment, all the rest as the atom data.
# Pawel Salek, 2006.
# Extend to convert Spartan "input" as well.
#
my %charge = ( 'H' => 1, 'C' => 6, 'N' => '7', 'O' => 8,
               'S' => 16 );

my $i;

die "Usage: xyz2mol.pl XYZ-FILE ...\n" unless $#ARGV >=0;
foreach $i (@ARGV) {
    unless(open FL, $i) {
	warn "cannot open $i for reading\n";
	next;
    }
    my $l1 = <FL>; chomp $l1;
    my $l2 = <FL>; chomp $l2;
    my $l3 = <FL>; chomp $l3;
    my ($totcharge, $spin) = split ' ', $l3;
    my %k = (); my %c = ();
    my $atoms = 0;
    while(<FL>){
        last if /^ENDCART/;
	s/^\s+//g;
	s/\s+$//g;
	($a) = ($_ =~ /^([a-zA-Z0-9]+)/);
        $k{$a} = '' unless defined $k{$a};
        $_ =~ s/^([a-zA-Z0-9]+)/$1  /g; # add spaces after atom name.
	$k{$a} = $k{$a} . "\n" . $_;
	$c{$a}++;
        $atoms++;
    }
    close FL;
    $ne = $i; $ne =~ s/\.xyz$/.mol/g;
    unless(open FL, ">$ne") {
	warn "cannot open $ne for writing.\n";
	next;
    }
    $at = scalar keys %c;
    print FL "ATOMBASIS\nLine1: $l1 Line2: $l2 Line3: $l3\n===========\n",
    "Atomtypes=$at Charge=$totcharge Angstrom\n";
    foreach $i (sort keys %k) {
        if($i>0) {
            $ch = 0+$i;
        } else {
            print "Atom '$i' unknown, file=$ne\n" unless  $charge{$i}>0;
            $ch = $charge{$i};
        }
	print FL "Charge=$ch Atoms=".$c{$i},
	" Basis=6-31G Aux=AhlrichsDenFit", $k{$i},"\n";
    }
    close FL;
    print "File: $i converted to $ne, $atoms atoms in total.\n";
}

File: LSDALTON/tools/xyz2mol.py
#!/usr/bin/python

#----------------------------------------------------------
# Title:   xyz2mol.py
# 
# Purpose: convert xyz file to DALTON MOLECULE.INP format
#          
# Usage:   $ ./xyz2mol.py filename.xyz
#          $ python xyz2mol.py filename.xyz
#
# Author:  Pablo Baudin
# Date:    Feb. 2015
#----------------------------------------------------------

import sys
import numpy as np

atoms=['H','He','Li','Be','B','C','N','O','F','Ne','Na','Mg','Al','Si','P','S','Cl','Ar',
       'K','Ca','Sc','Ti','V','Cr','Mn','Fe','Co','Ni','Cu','Zn','Ga','Ge','As','Se','Br',
       'Kr','Rb','Sr','Y','Zr','Nb','Mo','Tc','Ru','Rh','Pd','Ag','Cd','In','Sn','Sb','Te',
       'I','Xe',]

verbose=False

# Check that user provided an xyz file in input
if (len(sys.argv)!=2):
    print 'You need to provide only one argument, the .xyz file:\n'
    print '$ ./xyz2mol.py file.xyz'
    print '$ python xyz2mol.py file.xyz'
    exit()


# Get basis total charge and title
basis = raw_input('Enter basis set [cc-pVTZ]:\n')
if (basis==''):
    basis = 'cc-pVTZ\n'
else:
    basis+='\n'

totch = raw_input('Enter total charge of the molecule [0]:\n')
if (totch==''):
    totch = '0'

titl1 = raw_input('Enter title line 1 []:\n')
titl1+='\n'
titl2 = raw_input('Enter title line 2 []:\n')
titl2+='\n'

name = (sys.argv[1]).split('.')[0]

if (verbose):
    print '\nGenerating MOLECULE.INP file \n... \n'


# extract content of xyz file
xyzfile = open(sys.argv[1], "r")

lines = []
for line in xyzfile:
    lines.append(line)

xyzfile.close()

# Get number of atoms:
try:
    natoms = int(lines[0])
except:
    print 'file: '+sys.argv[1]+' does not fit the required format'
    exit()

# Check for title:
title=str(lines[1])
if (title!=''):
    if (titl1=='\n'):
        titl1=title
    elif (titl2=='\n'):
        titl2=title

coord=np.zeros((natoms,3))

o=2
labels=[]
Atomtypes=[]
for i in range(natoms):
    line = lines[i+o].split()
    atom = line[0]
    if (atom not in labels):
        Atomtypes.append(atom)

    labels.append(atom)

    for x in range(3):
        coord[i,x] = float(line[x+1])

# Print MOLECULE,INP file:
if (verbose):
    print 'Print MOLECULE.INP in terminal and file '+name+'.mol\n'
    print 'BASIS'
    print basis
    print titl1
    print titl2
    print 'Atomtypes='+str(len(Atomtypes))+' Nosymmetry Angstrom Charge='+totch

output = open(name+'.mol','w')
output.write('BASIS\n')
output.write(basis)
output.write(titl1)
output.write(titl2)
output.write('Atomtypes='+str(len(Atomtypes))+' Nosymmetry Angstrom Charge='+totch+'\n')

for atype in Atomtypes:
    try:
        charge=str(atoms.index(atype)+1)
        natomtype=str(labels.count(atype))
    except ValueError:
        print "Atom "+atype+" is not regognized by the script,"
        print "you might have to modify the script to include the full periodic table."
        # remove wrong output file:
        output.close()
        os.remove('./'+name+'.mol')
        exit()

    output.write('Charge='+charge+' Atoms='+natomtype+'\n')
    if (verbose):
        print 'Charge='+charge+' Atoms='+natomtype

    for i in range(natoms):
        if (labels[i] == atype):
            output.write((labels[i]+'  {0:10.6f}  {1:10.6f}  {2:10.6f} \n').format(coord[i,0],coord[i,1],coord[i,2]))
            if (verbose):
                print (labels[i]+'  {0:10.6f}  {1:10.6f}  {2:10.6f}').format(coord[i,0],coord[i,1],coord[i,2])

output.close()


print "(LS)DALTON molecule file: "+name+".mol has been created"


